Frontend/ci/prettier check

thdk opened the pull request:
PR to start running prettier check on CI pipelines to guarantee prettier has been run by the developer.

Not sure how I can test this, so hopefully I didn't mess up anything :crossed_fingers: 

Closes #1104

thdk committed:
Add prettier check to test stage of gitlab-ci

part of #1104

Idea is that the job should fail when the prettier still has formatting
issues.

Used npx so it doesn't needs a dependency on the build:frontend job for
prettier to be installed.

thdk committed:
add frontend build as dependency of prettier check

With frontend build as a dependency the /app folder will contain all
node modules including prettier.

So when running `npx prettier` it will look first if installed locally
and use that version.

darrenvong commented on app/.gitlab-ci.yml:
```diff
@@ -279,6 +279,21 @@ test:frontend:
+    default: false
+  script:
+    - cd /app
+    - npx prettier --check src/*
```
I think it's better to use the same version of Prettier already installed in the dev dependencies, since Prettier are pretty aggressive at changing the formatting rules between even non-major versions, so using `npx` will likely introduce discrepancies (since it tends to use the latest version) from what we got installed on the image.

lucaslcode replied:
(ie `yarn prettier --args`)

thdk replied:
> (ie `yarn prettier --args`)

aaah, I'm used to npm where `npm prettier --args` wouldn't work. But it seems yarn is smarter and knows we actually want to use a locally installed module.

I'll change it, thanks!

aapeliv commented on app/.gitlab-ci.yml:
```diff
@@ -279,6 +279,21 @@ test:frontend:
+  script:
+    - cd /app
+    - npx prettier --check src/*
+  rules:
+    - if: ($DO_CHECKS == "true") && ($CI_COMMIT_BRANCH == $RELEASE_BRANCH)
+      changes:
+      - app/frontend/**/*
```
This should probably be

```yaml
  rules:
    - if: ($DO_CHECKS == "true") && ($CI_COMMIT_BRANCH == $RELEASE_BRANCH)
      changes:
      - app/**/*
    - if: ($DO_CHECKS == "true") && ($CI_COMMIT_BRANCH != $RELEASE_BRANCH)
      changes:
      - app/proto/**/*
      - app/frontend/**/*
```

(Also `test:frontend` should probs be this too)

thdk replied:
> ```yaml
>   rules:
>     - if: ($DO_CHECKS == "true") && ($CI_COMMIT_BRANCH == $RELEASE_BRANCH)
>       changes:
>       - app/**/*
>     - if: ($DO_CHECKS == "true") && ($CI_COMMIT_BRANCH != $RELEASE_BRANCH)
>       changes:
>       - app/proto/**/*
>       - app/frontend/**/*
> ```

@aapeliv thanks for looking, I was indeed unsure about which rules should apply here.
Some questions to help me understand what is going on:

does this mean that for a
- commit on release branch (develop?)
  => run prettier check as soon as there is a change to a file that matches app/**/*
- commit not on release branch (PR branches, etc, ...)
 => run prettier check as soon as there is a change to a file that matches either app/proto/ or app/frontend/

If so, why do we need to run prettier on a release branch even if nothing changed to files that should have been formatted with prettier. Is it just to be sure because it is the release branch?

Why include proto folder for non release branches? I don't thinks these changes could affect formatting of frontend. But that's just my first idea.

I do think including proto folder for tests does make sense.

thdk replied:
I've updated the rules as you suggested for test:frontend.
For test:frontend-prettier, I've done the same except that I left out the `protos` folder. Let me know if that line would still be required for this PR.

aapeliv reviewed (commented):
Looks good with or without comment

thdk committed:
use yarn instead of npx

thdk committed:
Update gitlab job rules for test:frontend and test:frontend-prettier

thdk committed:
Revert last commit because it breaks yaml

This reverts commit de3bd40fef9613182ed5305d48f2a2fb60910ef0.

thdk committed:
add rule for branches other than release branch

thdk committed:
add rule for non release branch to build:frontend

This in attempt to fix gitlab lint errors:

'test:frontend-prettier' job needs 'build:frontend' job
but 'build:frontend' is not in any previous stage

thdk commented:
Ah now it works and the prettier job fails because of formatting errors. Hooray! 🎉 

```
Downloading artifacts for build:frontend (1524910834)...
Downloading artifacts from coordinator... ok        id=1524910834 responseStatus=200 OK token=yUSjuSXf
Executing "step_script" stage of the job script
00:18
Using docker image sha256:ef095efae9fc6703400198b5a0f4cbb59961803a9ace4eb0bc85cdaf9cb7e267 for registry.gitlab.com/couchers/couchers/frontend:frontend-ci-prettier-check-44f4b368 with digest registry.gitlab.com/couchers/couchers/frontend@sha256:51ddd8b4a599d2dfffc35f5e8e573cfda99e5f007d063db6f7b3e60cb575f9a1 ...
$ cd /app
$ yarn prettier --check src/*
yarn run v1.22.5
$ /app/node_modules/.bin/prettier --check src/App.css src/App.tsx src/AppRoute.tsx src/AppRoutes.tsx src/components src/constants.ts src/features src/index.css src/index.tsx src/proto src/queryKeys.ts src/react-app-env.d.ts src/react-lines-ellipsis.d.ts src/reactQueryClient.tsx src/resources src/routes.ts src/service src/serviceWorker.ts src/setupTests.ts src/stories src/test src/theme.ts src/toast-ui-viewer.d.ts src/utils
Checking formatting...
[warn] src/components/Button/Button.tsx
[warn] src/components/ContributorForm/StandaloneContributorForm.tsx
[warn] src/features/communities/CommunityBase.tsx
[warn] src/features/communities/discussions/DiscussionPage.tsx
[warn] src/features/communities/GroupPage.tsx
[warn] src/features/communities/PagePage.tsx
[warn] src/features/connections/friends/FriendRequestsReceived.stories.tsx
[warn] src/features/connections/friends/FriendRequestsSent.stories.tsx
[warn] src/features/messages/groupchats/InviteDialog.tsx
[warn] src/features/profile/actions/MessageUserButton.test.tsx
[warn] src/features/profile/view/leaveReference/formSteps/Appropriate.tsx
[warn] src/features/profile/view/leaveReference/formSteps/Rating.tsx
[warn] src/features/profile/view/leaveReference/formSteps/submit/SubmitReference.tsx
[warn] src/features/profile/view/leaveReference/formSteps/Text.tsx
[warn] src/features/profile/view/leaveReference/LeaveReferencePage.tsx
[warn] src/features/profile/view/UserPage.tsx
[warn] src/features/search/SearchPage.tsx
[warn] src/features/user/UserTabContext.tsx
[warn] src/utils/hooks.ts
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Cleaning up file based variables
00:01
ERROR: Job failed: exit code 1
```

thdk committed:
Merge branch 'develop' into frontend/ci/prettier-check

thdk committed:
add rule for non release branch to test:frontend

aapeliv reviewed (approved):
LGTM
